src/main/scala/tech/sourced/gemini/Gemini.scala

```diff
@@ -81,39 +83,120 @@ class Gemini(session: SparkSession, log: Slf4jLogger, keyspace: String = Gemini.
-      val features = extractFeatures(feClient, uast.get)
-      log.info(s"features: ${features.toString}")
-      // TODO: calculate minhash, find similar
+  def findSimilarForFile(
+                          file: File,
+                          conn: Session,
+                          bblfshClient: BblfshClient,
+                          feClient: FeatureExtractor,
+                          docFreqFile: String,
+                          paramsFile: String,
+                          htnum: Int,
+                          bandSize: Int): Option[Array[String]] = {
+
+    val features = extractUAST(file, bblfshClient).map { uast =>
+      log.debug(s"uast received: ${uast.toString}")
+
+      val result = extractFeatures(feClient, uast)
+      log.debug(s"features: ${result.toString}")
+
+      result
+    }
+
+    features match {
+      case Some(featuresList) if featuresList.isEmpty => {
+        log.warn("file doesn't contain features")
+        None
+      }
+      case Some(featuresList) => {
+        val wmh = hashFile(featuresList, new File(docFreqFile), new File(paramsFile))
+
+        // split hash to bands
+        // a little bit verbose because java doesn't have uint32 type
+        // in apollo it's one-liner:
+        // https://github.com/src-d/apollo/blob/57e52394783d73e38cf1f862afc0166724991fd5/apollo/query.py#L35
+        val bands = (0 until htnum).map { i =>
+          val from = i * bandSize
+          val to = (i + 1) * bandSize
+          val band = (from until to).foldLeft(Array.empty[Byte]) { (arr, j) =>
+            arr ++ MathUtil.toUInt32ByteArray(wmh(j)(0)) ++ MathUtil.toUInt32ByteArray(wmh(j)(1))
+          }
+          band
+        }
+
+        log.info("Looking for similar items")
+        val similar = bands.zipWithIndex.foldLeft(Set.empty[String]) { case (similar, (band, i)) =>
+          val table = "hashtables"
+          val cql = s"SELECT sha1 FROM $keyspace.$table WHERE hashtable=$i AND value=0x${MathUtil.bytes2hex(band)}"
+          log.debug(cql)
+
+          val sha1s = conn.execute(new SimpleStatement(cql))
+            .asScala
+            .map(row => row.getString("sha1"))
+
+          similar ++ sha1s
+        }
+        log.info(s"Fetched ${similar.size} items")
+
+        Some(similar.toArray)
+      }
+      case _ => None
     }
   }
 
   def extractUAST(file: File, client: BblfshClient): Option[Node] = {
     val byteArray = Files.readAllBytes(file.toPath)
 
-    val resp = client.parse(file.getName, new String(byteArray))
-    if (resp.errors.nonEmpty) {
-      val errors = resp.errors.mkString(",")
-      log.error(s"bblfsh errors: ${errors}")
-    }
+    try {
+      val resp = client.parse(file.getName, new String(byteArray))
+      if (resp.errors.nonEmpty) {
+        val errors = resp.errors.mkString(",")
+        log.error(s"bblfsh errors: ${errors}")
+      }
 
-    resp.uast
+      resp.uast
+    } catch {
+      case e: Exception => {
```

Comment:
[`NonFatal(e)`](http://www.scala-lang.org/api/current/scala/util/control/NonFatal$.html) would be more appropriate than `e: Exception`.